### PR TITLE
ENH: spatial.transform.Rotation.random [GSOC 2018]

### PR DESCRIPTION
**Pull request by adbugger**
Uniform random sampling of 3D rotations

### CODE REVIEW

**Comment by adbugger** [comment]
@nmayorov here is the prepared pull request based off the latest master

File: scipy/spatial/transform/rotation.py [hunk 1]
```diff
@@ -1460,3 +1461,55 @@ def __getitem__(self, indexer):
...
+        array([[ 0.43748455,  0.23734269, -0.80606684,  0.32020612]])
+        >>> R.random(5).as_quat()
+        array([[ 0.35598949,  0.62048719, -0.69368101, -0.08410586],
+               [-0.62061285, -0.75545239,  0.15646323, -0.14018072],
+               [-0.31340545,  0.55852902, -0.13422706,  0.75617819],
+               [ 0.25086769, -0.37389821, -0.53738335,  0.71308111],
+               [ 0.78349346, -0.03642139, -0.53405077,  0.31559667]])
+       """
+        if random_state is None:
+            random_state = np.random.RandomState(seed=None)
```
Comment (rkern) [inline scipy/spatial/transform/rotation.py]: The default should be the global `RandomState` instance, not a new one.

https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/validation.py#L786-L804

Comment (rkern) [reply scipy/spatial/transform/rotation.py]: Since you aren't going to store the `random_state` anywhere, a handy shortcut is to do

```
if random_state is None:
    random_state = np.random
```

Comment (andyfaff) [reply scipy/spatial/transform/rotation.py]: Probably use the `scipy._util.check_random_state` function, it's copied and pasted from scikit-learn

File: scipy/spatial/transform/rotation.py [hunk 2]
```diff
@@ -1460,3 +1462,54 @@ def __getitem__(self, indexer):
...
         """
         return self.__class__(self._quat[indexer], normalized=True)
+
+    @classmethod
+    def random(cls, num=None, random_state=None):
+        """Generate uniformly distributed rotations.
+
+        Parameters
+        ----------
+        num : None or int
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: Both optional, to be pedantic.

File: scipy/spatial/transform/rotation.py [hunk 3]
```diff
@@ -1460,3 +1462,54 @@ def __getitem__(self, indexer):
...
+            Number of random rotations to generate. If None, then a single
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: Alternative wording I like "if None (default), ...". But you surely can keep your variant if you prefer.

File: scipy/spatial/transform/rotation.py [hunk 4]
```diff
@@ -1460,3 +1462,54 @@ def __getitem__(self, indexer):
...
+
+        Examples
+        --------
+        >>> from scipy.spatial.transform import Rotation as R
+
+        Sample a single rotation:
+
+        >>> R.random().as_quat()
+        array([-0.07838344, -0.25840318, -0.24463331, -0.93125634])
+        >>> R.random(None).as_quat()
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: This example looks unnecessary.

File: scipy/spatial/transform/rotation.py [hunk 5]
```diff
@@ -1460,3 +1462,54 @@ def __getitem__(self, indexer):
...
+        array([-0.24475285,  0.58126954, -0.54600616,  0.55145176])
+
+        Sample a stack of rotations:
+
+        >>> R.random(1).as_quat()
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: I think this is marginally useful as well.

File: scipy/spatial/transform/rotation.py [hunk 6]
```diff
@@ -1460,3 +1462,54 @@ def __getitem__(self, indexer):
...
+        array([[ 0.43748455,  0.23734269, -0.80606684,  0.32020612]])
+        >>> R.random(5).as_quat()
+        array([[ 0.35598949,  0.62048719, -0.69368101, -0.08410586],
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: What if we use Euler angles instead? To roughly demonstrate "uniformity".

**Review by nmayorov — COMMENTED** [review]
Will merge after you consider my minor suggestions.

File: scipy/spatial/transform/rotation.py [hunk 7]
```diff
@@ -1460,3 +1462,54 @@ def __getitem__(self, indexer):
...
+    def random(cls, num=None, random_state=None):
+        """Generate uniformly distributed rotations.
+
+        Parameters
+        ----------
+        num : None or int
+            Number of random rotations to generate. If None, then a single
+            rotation is generated. Default is None.
+        random_state : None, int or `numpy.random.RandomState` object
+            If RadomState object, used for generator state. If integer, used
```
Comment (ilayn) [inline scipy/spatial/transform/rotation.py]: Typo `RandomState`. 

Also you can simply write `None, optional` as the type and explain the possibilities in the description. E.g., 

    It accepts an `int` as a seed for the random generator or a RandomState object. 

Then you can explain what else this keyword accepts.

**Comment by ilayn** [comment]
Not to step on toes, but do we really need to cut the word `as_quat` instead of `as_quaternion` and its `from_<>` counterpat? 

It feels a bit matlabish and not necessarily offers a great deal of keystroke economy.

**Comment by nmayorov** [comment]
We totally don't. I'm slightly in favor of `as_quat`, because we already use some acronyms and it feel natural and unambiguous to me.

@adbugger @larsoner @rgommers can you also add your votes? I'm +0.25 to `as_quat`.

File: scipy/spatial/transform/rotation.py (hunk 7 above)
Comment (rkern) [reply scipy/spatial/transform/rotation.py]: Bikeshedding nitpick: with optional arguments, you should describe the type as what you intend people to actually pass when they pass something rather than the type of the default when they pass nothing. Personally, I'd write this as `random_state : RandomState or int, optional`

Comment (rkern) [reply scipy/spatial/transform/rotation.py]: Or keep the current wording and tack on `, optional`. E.g. https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/gaussian_process/gpr.py#L108

**Comment by rgommers** [comment]
No strong opinion. I'd say I'm -0.25; mainly because I dislike things that smell like Matlab.

File: scipy/spatial/transform/rotation.py (hunk 6 above)
Comment (adbugger) [reply scipy/spatial/transform/rotation.py]: I'm not sure if that is an effective demonstration of uniformity, but still a better suggestion than quaternions.

**Comment by adbugger** [comment]
I had initially named it as `_quaternion` itself. I felt that the other formats are widely known enough to merit abbreviation. Using the full form here conveys more information at the cost of very few keystrokes.

+0.5 for changing it back to `_quaternion`. Should I open another PR for that?

**Comment by nmayorov** [comment]
Rebase on the current master then I'm merging.

For renaming --- let's wait some time.

**Comment by adbugger** [comment]
Rebasing alone does not work. Resolving conflicts and squashing commit history.

**Comment by nmayorov** [comment]
Could you try to do rebase? As far as I understand it should not be different from merging, you resolve the conflicts the same way.

**Comment by adbugger** [comment]
So rebasing from command line worked.

**Comment by adbugger** [comment]
Can this be merged?

**Comment by nmayorov** [comment]
Sure, just forgot about it.
